src/plugins/interactive_setup/server/plugin.ts

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { schema } from '@kbn/config-schema';
+import type { CorePreboot, PrebootPlugin, PluginInitializerContext } from 'src/core/server';
+import fs from 'fs/promises';
+import { errors } from '@elastic/elasticsearch';
+import Boom from '@hapi/boom';
+import chalk from 'chalk';
+import type { ConfigType } from './config';
+
+export class InteractiveSetupPlugin implements PrebootPlugin {
+  readonly #initializerContext: PluginInitializerContext<ConfigType>;
+  constructor(initializerContext: PluginInitializerContext<ConfigType>) {
+    this.#initializerContext = initializerContext;
+  }
+
+  public setup(core: CorePreboot) {
+    const { enabled } = this.#initializerContext.config.get<ConfigType>();
+
+    const isDev = this.#initializerContext.env.mode.dev;
+    const isConfigured =
+      core.elasticsearch.config.credentialsSpecified && core.elasticsearch.config.hosts.length;
+
+    /**
+     * env: dev -> enabled: false -> skip
+     * env: dev -> enabled: undefined -> skip
+     * env: dev -> enabled: true -> hold setup
+     * env: prod -> enabled: false -> skip
+     * env: prod -> enabled: undefined -> hold setup if not configured
+     * env: prod -> enabled: true -> hold setup if not configured
+     */
+    const shouldHoldSetupPhase = isDev
+      ? enabled === true
+      : enabled !== false
+      ? !isConfigured
+      : false;
+
+    if (!shouldHoldSetupPhase) {
+      return;
+    }
+
+    core.http.registerRoutes('', (prebootRouter) => {
+      let completeSetup: (result: { shouldReloadConfig: boolean }) => void;
+
+      prebootRouter.post(
+        {
+          path: '/api/preboot/setup',
+          validate: {
+            body: schema.oneOf([
+              schema.object({
+                hosts: schema.arrayOf(schema.string(), { minSize: 1 }),
+                apiKey: schema.string(),
+                caFingerprint: schema.string(),
+              }),
+              schema.object({
+                hosts: schema.arrayOf(schema.string(), { minSize: 1 }),
+                username: schema.string(),
+                password: schema.string(),
+                caFingerprint: schema.string(),
+              }),
+            ]),
+          },
+          options: { authRequired: false },
+        },
+        async (context, request, response) => {
+          if (!core.preboot.isSetupOnHold()) {
```

Comment:
__note:__ it'd make sense to log every such condition, and, in general, we may want to log as much as feasible :slightly_smiling_face: